keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1342 @@
+
+  # Send to random endpoint
+  random mod 2 vmap {
+    0 : dnat to 10.180.0.4:8000
+    1 : dnat to 10.180.0.5:8000
+  }
+}
+```
+
+With the current set of nftables functionality, it does not seem
+possible to do this (in the case where affinity is in use), but future
+features may make it possible.
+
+It is not yet clear what the tradeoffs of such rewrites are, either in
+terms of runtime performance, or of admin/developer-comprehensibility
+of the ruleset.
+
+[kpng nftables backend]: https://github.com/kubernetes-sigs/kpng/tree/master/backends/nft
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+The goal is to ensure that we don't accept enhancements with inadequate testing.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+<!--
+Based on reviewers feedback describe what additional tests need to be added prior
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+##### Unit tests
+
+We will add unit tests for the `nftables` mode that are equivalent to
+the ones for the `iptables` mode. In particular, we will port over the
+tests that feed Services and EndpointSlices into the proxy engine,
+dump the generated ruleset, and then mock running packets through the
+ruleset to determine how they would behave.
+
+The `cmd/kube-proxy/app` tests mostly only test configuration parsing,
+and we will extend them to understand the new mode and its associated
+configuration options, but there will not be many changes made there.
+
+<!--
+Additionally, for Alpha try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+The data can be easily read from:
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `<package>`: `<date>` - `<test coverage>`
+
+##### Integration tests
+
+Kube-proxy does not have integration tests.
+
+##### e2e tests
+
+Most of the e2e testing of kube-proxy is backend-agnostic. Initially,
+we will need a separate e2e job to test the nftables mode (like we do
+with ipvs). Eventually, if nftables becomes the default, then this
+would be flipped around to having a legacy "iptables" job.
+
+The handful of e2e tests that specifically examine iptables rules will
```

Comment:
I don't know if this still exist, but is not great for e2e test to have this implementation dependencies, I rather rewrite them to focus on the expected behavior or drop them entirely, I think you already did that for some of them

Reply:
Interesting... I had thought there were more of these. I guess they mostly got fixed. (You're right, [I did just remove one of them](https://github.com/kubernetes/kubernetes/pull/113335/commits/c6cc056675314ef7f74e4b318d39eadd73b2de40).) There is only one remaining test that parses iptables rules, [`"It should recreate its iptables rules if they are deleted [Disruptive]"`](https://github.com/kubernetes/kubernetes/blob/v1.26.1/test/e2e/network/networking.go#L551), which tests 2 things:
1. kubelet recreates `KUBELET-IPTABLES-HINT` if it is deleted (which is iptables-specific and should stick around regardless of kube-proxy, for as long as we expect third-party components to use iptables).
2. Services eventually start working again even if you delete all `KUBE-*` iptables chains. It doesn't actually test that they _stop_ working after you delete those chains, so it would still pass in nftables mode, it just wouldn't test anything interesting. (And as discussed in the KEP text, we probably don't need to do monitoring for nftables.)

There are also a few tests that manually invoke iptables to break the network via [`TestUnderTemporaryNetworkFailure`](https://github.com/kubernetes/kubernetes/blob/v1.27.0-alpha.2/test/e2e/framework/network/utils.go#L1078) (plus [one that doesn't use that function but should](https://github.com/kubernetes/kubernetes/blob/v1.27.0-alpha.2/test/e2e/apimachinery/etcd_failure.go#L63)). This will eventually need to be updated to have an nftables-based version, but that's independent of kube-proxy.

I think that's it? Yay us.

Reply:
filed https://github.com/kubernetes/kubernetes/issues/119047 about TestUnderTemporaryNetworkFailure